Improved the Product controller class

Sethbass opened the pull request:
<!-- Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows with the necessary information: -->

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Code improvements and factoring.
| Type?         | bug fix
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | http://forge.prestashop.com/browse/BOOM-2950
| How to test?  | Test the Product pages, tests must pass.

<!-- Click the form's "Preview button" to make sure the table is functional in GitHub. Thank you! -->

#### Important guidelines

* Make sure [your local branch is up to date](https://help.github.com/articles/syncing-a-fork/) before commiting your changes!
* Your code MUST respect [our Coding Standards](http://doc.prestashop.com/display/PS16/Coding+Standards) (for code written in PHP, JavaScript, HTML/CSS/Smarty/Twig, SQL)!
* Your commit name MUST respect our [naming convention](http://doc.prestashop.com/display/PS16/How+to+write+a+commit+message)!

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/8677)
<!-- Reviewable:end -->

Commit:
Changes to improve the product controller and enable more than 62 combinations

Commit:
Edit the test and check PSR2

mickaelandrieu commented:
Hello @Sethbass, I like a lot your refactoring *but* allow CSRF in Product page is a no go for obvious reasons :)

Can you detail which error(s) is returned in your case? I've already worked with a form with more than 150 fields and I can tell this is not an issue, but maybe your `php_value post_max_size` is not big enough for the product form and needs to be increased.

So, for the first step if you remove the `csrf_protection => false`, for me it's :+1:  and mergeable. We can discuss later on how we can fix your pretty annoying issue :)

And thanks for this refactoring! Good job!

mickaelandrieu commented on src/PrestaShopBundle/Controller/Admin/ProductController.php:
```diff
@@ -620,6 +608,45 @@ public function formAction($id, Request $request)
+        $combinations = $product->getAttributesResume(
+            $this->get('prestashop.adapter.legacy.context')->getContext()->language->id
+        );
+        if (is_array($combinations)) {
```
```diff
- is_array($combinations)
+ is_array($combinations) && count($combinations) > 0
```

Quetzacoalt91 replied:
If we got an error, I'd have approved your answer.
But is it really important? In the case `count() === 0`, the foreach() won't loop but exit immediately.

mickaelandrieu replied:
We don't write the code only for the PHP engine, it's only me but It's easier to read and understand what is done here. The best would be to check if $combinations is always an array (I think it is) and only keep the check on the count, wdyt?

Commit:
BO: Several improvements and refactoring in the product Controller

Commit:
BO: Several improvements and refactoring in the product Controller

Commit:
BO: Several improvements and refactoring in the product Controller

mickaelandrieu commented:
@Sethbass would you mind to allow the maintainers to contribute? You need to check the checkbox on the right under participants section. I'd like to contribute on your contribution :+1:

Sethbass commented:
Hi Michael,

It looks like it is already checked ;)

So just to explain a bit how I got there : I was facing issues with the updates of a product with like 500 combinations (customer project) on 1.7.2.4. The initial error was "CRSF token is invalid, so I removed the CSRF validation just to see if it was the only issue. Then it passed, meaning green alert on the right saying that my changes were saved. Unfortunately that was not the case, pressing the save button and refreshing the page, my changes disappeared. So I have decided to dig a bit further and after many hours of debugging I understood that few inputs were disappearing randomly from the request giving me unexpected behavior and a big headache ;) I ended up finding out that it was linked to the PHP settings max_input_vars that is set to 1000 by default. I changed the value to 10000 and it solved all my issues.
So remains the issue with the CSRF token, testing it this morning I found out that with the max_input_vars set to 10000 it was also solving this issue.

As you can imagine we cannot leave it as it, because the application needs to work with a standard PHP version and having a form with thousands of fields will just make things impossible with Doctrine behind the scene. This page is already bugging on my mac (brand new with core i7 and 32 Go + SSD)... 
My idea was to move the combinations on a dedicated page => not the same controller and to use pagination (50 per page would be good). Then to allow the edition of the combinations one by one and not altogether. Another feature that I would like to add is the possibility to filter them with attribute picklists so that would be easier to work with. The good thing is also that it will ease the move to doctrine and help us to put the right validation on the form using entity Assert VS asserts in the form. 

Could you please tell me what your thoughts about it?

So before doing anything, I needed to understand what was done in this controller and make it easier so we can work in a good way.

Moving this to doctrine sounds to me like a huge priority as it would remove the usage of previous controllers. Most of the services we have here are just doing doctrine's job and make things extremely complicated to understand.

Any good idea is more than welcome.

Cheers,

Seb

Commit:
BO: Several improvements and refactoring in the product Controller

Sethbass commented:
@mickaelandrieu : I'm about to start the creation of the product entity and repository. The idea is to start to use it on this Product controller. Did anyone start that job? Is that a good idea?

Thanks a lot,

Seb

mickaelandrieu commented:
> I changed the value to 10000 and it solved all my issues.
So remains the issue with the CSRF token, testing it this morning I found out that with the max_input_vars set to 10000 it was also solving this issue.

:+1:  but I think it's ok to have to setup this value in case of product with a lot of combinations! To be honest have 500 combinations for a single product is very uncommon...

> My idea was to move the combinations on a dedicated page => not the same controller and to use pagination (50 per page would be good). 

I'm not a Product Owner so I can't tell, but I'm in favor of trying to improve the performance of this page first. From UX point of view, I'm not sure.

> Moving this to doctrine sounds to me like a huge priority as it would remove the usage of previous controllers.

It's in the roadmap for the 1.7.x release. As you can see, the Product Page is not really migrated to Symfony for now. This was the very first page and a lot of mistakes have been done! We are now more easy with the Symfony framework. Our plan here is:
1. remove calls from the old legacy controller
2. use a doctrine repository but without related entity (as we need to think about our model, but afaik it's out of scope of 1.7.x): you can see what I mean regarding how we have migrated the Performance Page for instance;
3. So, to sum up, I'll try to help you on this really big challenge :+1: but let's start little to get a chance to see this merged :)

mickaelandrieu committed:
Extracted private functions into dedicated services

mickaelandrieu committed:
Extracted Product Categories form outside of Controller

mickaelandrieu commented:
So, there is still work to do but the ProductController is now under 900 lines of code, comments included :trollface:

eternoendless commented:
Reviewed 1 of 3 files at r1, 1 of 2 files at r2, 1 of 2 files at r3, 10 of 10 files at r5.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[src/Adapter/Product/FilterCategoriesRequestPurifier.php, line 46 at r5](https://reviewable.io:443/reviews/prestashop/prestashop/8677#-L3i51k9cY4jg3nDIW1W:-L3i51kAaToGHQp3JW5v:bsdofn0) ([raw file](https://github.com/prestashop/prestashop/blob/6aee5703fce41c831032c2d572f0cf0f841c4a61/src/Adapter/Product/FilterCategoriesRequestPurifier.php#L46)):*
> ```PHP
>                 if (
>                     $param === 'filter_category' &&
>                     (!is_numeric($value) || (is_numeric($value) && $value < 0))
> ```

`$value` is either numeric or not, so this can be simplified:

```php
(!is_numeric($value) || $value < 0)
```

---

*[src/Adapter/Product/FilterParametersUpdater.php, line 70 at r5](https://reviewable.io:443/reviews/prestashop/prestashop/8677#-L3lz27q708HwAXEZsL7:-L3lz27q708HwAXEZsL8:brqxjml) ([raw file](https://github.com/prestashop/prestashop/blob/6aee5703fce41c831032c2d572f0cf0f841c4a61/src/Adapter/Product/FilterParametersUpdater.php#L70)):*
> ```PHP
>         &$orderBy,
>         &$sortOrder
>     ) {
> ```

Passing parameters by reference is not generally not a good idea unless you have a very specific reason to do so.

I think this should be broken into 4 different methods like this:

```php
public function getOffset($offset, $filterParameters)
{
    return ($offset === 'last' && isset($filterParameters['last_offset'])) ? $filterParameters['last_offset'] : $offset;
}
```

---

*[src/PrestaShopBundle/Controller/Admin/CommonController.php, line 212 at r5](https://reviewable.io:443/reviews/prestashop/prestashop/8677#-L3i5Fm1U_SkLzXog-gt:-L3i5Fm1U_SkLzXog-gu:b-lb7a05) ([raw file](https://github.com/prestashop/prestashop/blob/6aee5703fce41c831032c2d572f0cf0f841c4a61/src/PrestaShopBundle/Controller/Admin/CommonController.php#L212)):*
> ```PHP
> 
>     /**
>      * Specific action to render a specific field multiple twice.
> ```

Small error here: "multiple twice" :)

---

*[src/PrestaShopBundle/Controller/Admin/CommonController.php, line 221 at r5](https://reviewable.io:443/reviews/prestashop/prestashop/8677#-L3mPJ8QreHZSMFF8t0f:-L3mOo6FP2De3MG8_EiK:bvtj2iz) ([raw file](https://github.com/prestashop/prestashop/blob/6aee5703fce41c831032c2d572f0cf0f841c4a61/src/PrestaShopBundle/Controller/Admin/CommonController.php#L221)):*
> ```PHP
>      * @return Response
>      */
>     public function renderFieldAction($formName, $formType, $fieldName, $fieldData)
> ```

Why put this here? Isn't this specific to the Product page?

---

*[src/PrestaShopBundle/Controller/Admin/ProductController.php, line 139 at r5](https://reviewable.io:443/reviews/prestashop/prestashop/8677#-L3m1AYG-2ZVD041eN-7:-L3m1AYG-2ZVD041eN-8:b-rmxrs3) ([raw file](https://github.com/prestashop/prestashop/blob/6aee5703fce41c831032c2d572f0cf0f841c4a61/src/PrestaShopBundle/Controller/Admin/ProductController.php#L139)):*
> ```PHP
>         }
> 
>         $persistedFilterParameters = $this->get('prestashop.adapter.filter_parameters_updater')
> ```

You could reuse the previous instance of this class

---

*[src/PrestaShopBundle/Controller/Admin/ProductController.php, line 941 at r5](https://reviewable.io:443/reviews/prestashop/prestashop/8677#-L3m3t-UNqyV0vPhhp-j:-L3m3t-UNqyV0vPhhp-k:bdibbel) ([raw file](https://github.com/prestashop/prestashop/blob/e651445a3436b69f4b2e1a45062b5081ecd4789e/src/PrestaShopBundle/Controller/Admin/ProductController.php#L941)):*
> ```PHP
>         /* @var $urlGenerator UrlGeneratorInterface */
>         return $this->redirect($urlGenerator->generate('admin_product_catalog'));
>     }
> ```

This is a Backward compatibility (BC) break that should be noted in the PR description

---

*[src/PrestaShopBundle/Model/Product/AdminModelAdapter.php, line 435 at r5](https://reviewable.io:443/reviews/prestashop/prestashop/8677#-L3mQWWiR74-sPPQN7t1:-L3mQWWiR74-sPPQN7t2:b70ifk9) ([raw file](https://github.com/prestashop/prestashop/blob/6aee5703fce41c831032c2d572f0cf0f841c4a61/src/PrestaShopBundle/Model/Product/AdminModelAdapter.php#L435)):*
> ```PHP
>         $this->mapStep4FormData($product);
>         $this->mapStep5FormData($product);
>         $this->mapStep6FormData($product);
> ```

IMO it's better to make assignments explicit instead of burying it in the method:

```php
$this->formData['step1'] = $this->getStep1FormData($product);
$this->formData['step2'] = $this->getStep2FormData($product);
// ...
```

I believe this way the code is easier to read

---

*[src/PrestaShopBundle/Model/Product/AdminModelAdapter.php, line 529 at r5](https://reviewable.io:443/reviews/prestashop/prestashop/8677#-L3mRiWIZf1KQSJShBkH:-L3mRiWIZf1KQSJShBkI:b9i5dvf) ([raw file](https://github.com/prestashop/prestashop/blob/e651445a3436b69f4b2e1a45062b5081ecd4789e/src/PrestaShopBundle/Model/Product/AdminModelAdapter.php#L529)):*
> ```PHP
>     }
> 
>     public function getAttributesResume()
> ```

Where did this method go?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8677)*
<!-- Sent from Reviewable.io -->

mickaelandrieu commented:
@eternoendless this is really difficult to understand your review using Github, is there a way to put your comments in the "right" places?

mickaelandrieu commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[src/PrestaShopBundle/Controller/Admin/CommonController.php, line 221 at r5](https://reviewable.io:443/reviews/prestashop/prestashop/8677#-L3mPJ8QreHZSMFF8t0f:-L3suWe24Gx5vnsbnUE4:bhysthi) ([raw file](https://github.com/prestashop/prestashop/blob/6aee5703fce41c831032c2d572f0cf0f841c4a61/src/PrestaShopBundle/Controller/Admin/CommonController.php#L221)):*
<details><summary><i>Previously, eternoendless (Pablo Borowicz) wrote…</i></summary><blockquote>

Why put this here? Isn't this specific to the Product page?
</blockquote></details>

not anymore and its really more performant! But I need to re-introduce the old once and mark it as deprecated for BC:+1

---

*[src/PrestaShopBundle/Controller/Admin/ProductController.php, line 139 at r5](https://reviewable.io:443/reviews/prestashop/prestashop/8677#-L3m1AYG-2ZVD041eN-7:-L3sv1fV2gFaBB82D-Yi:b-frt6lx) ([raw file](https://github.com/prestashop/prestashop/blob/6aee5703fce41c831032c2d572f0cf0f841c4a61/src/PrestaShopBundle/Controller/Admin/ProductController.php#L139)):*
<details><summary><i>Previously, eternoendless (Pablo Borowicz) wrote…</i></summary><blockquote>

You could reuse the previous instance of this class
</blockquote></details>

I dont see the code you refers to.

---

*[src/PrestaShopBundle/Controller/Admin/ProductController.php, line 941 at r5](https://reviewable.io:443/reviews/prestashop/prestashop/8677#-L3m3t-UNqyV0vPhhp-j:-L3sv6_daVQ1rSVWqxYv:b-s93mln) ([raw file](https://github.com/prestashop/prestashop/blob/e651445a3436b69f4b2e1a45062b5081ecd4789e/src/PrestaShopBundle/Controller/Admin/ProductController.php#L941)):*
<details><summary><i>Previously, eternoendless (Pablo Borowicz) wrote…</i></summary><blockquote>

This is a Backward compatibility (BC) break that should be noted in the PR description
</blockquote></details>

I'll do something backward compatible instead

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8677)*
<!-- Sent from Reviewable.io -->

mickaelandrieu commented on src/PrestaShopBundle/Resources/config/admin/routing_product.yml:
```diff
@@ -60,14 +60,6 @@ admin_product_form:
     requirements:
         id: \d+
 
-admin_product_use_legacy_setter:
```
this is not a BC, this code and the related code should be deleted before 1.7.0!

mickaelandrieu commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[src/PrestaShopBundle/Controller/Admin/ProductController.php, line 941 at r5](https://reviewable.io:443/reviews/prestashop/prestashop/8677#-L3m3t-UNqyV0vPhhp-j:-L3trDgMzZhnHGTn_Xwt:bgy70cx) ([raw file](https://github.com/prestashop/prestashop/blob/e651445a3436b69f4b2e1a45062b5081ecd4789e/src/PrestaShopBundle/Controller/Admin/ProductController.php#L941)):*
<details><summary><i>Previously, mickaelandrieu (Mickaël Andrieu) wrote…</i></summary><blockquote>

I'll do something backward compatible instead
</blockquote></details>

erratum: I unremove renderFieldAction but not this one, this is something that should been removed before 1.7.0 release (previously, you could access the old product pages from 1.6). This pr doesn't introduce any "new" BCs.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8677)*
<!-- Sent from Reviewable.io -->

mickaelandrieu committed:
Re-introduct ProductController:renderField action (BC)

mickaelandrieu committed:
Address comments on FilterCategoriesRequestPurifier class

mickaelandrieu commented:
> I think this should be broken into 4 different methods like this:

I'd like to avoid the getter/setter options for everything. I rework the function.

mickaelandrieu committed:
Extracted export action into specific service

mickaelandrieu committed:
Used trans helper instead of calling translator

mickaelandrieu committed:
Fixed coding style issues

mickaelandrieu committed:
Introduced getAdminLink helper

mickaelandrieu commented:
# `catalogAction` explained

This action is in charge of rendering the Product list page, except for the list of products rendered by `listAction`. This controller has so (too?!) many responsibilities:

1. Checks the user permissions
2. Set the user locale in session
3. Clean the filters coming from the request
4. Clean again the filters coming from the request
5. Get products from the request filters
6. Store the last SQL query for the "Export|SQL manager" actions
7. If no products are found => render a different template (catalog empty page)
8. If products are found => create the product categories form
9. Guess what? one more action on filters coming from the request
10. Retrieve a permission error if any
11. Some checks for drag & drop feature...

## What I suggest

1. Task 1 can be done before entering the controller thanks to *KernelEvents::CONTROLLER* event. More, this can be re-used for almost every action of Symfony pages.
2. Task 2 is useless as it's done by default by the `UserLocaleListener` => To be removed
3. Task 3, 4 and 9 should be done once and for all using a new Immutable object call *ProductsFilters* that we retrieve directly as catalog Action arguments. This may change the behavior in case of Task 9 but remember, the list of products is retrieved later using list actions...
4. Task 5 and 6 can stay in `catalogAction` but improved. Something like:
```php
list($products, $lastQuery) = $this->get('prestashop.products_provider')->retrieveProductsAndQuery($filters);
```
Make so much sense here: more, the service should be immutable and for me it feel so wrong to store the last executed SQL query inside the service!

5. Task 7 should be handled by a specific action and can be done before the controller thanks to *KernelEvents::CONTROLLER* event. This listener is - of course - specific to the *ProductController*.
6. Task 8 should of course stay in `catalogAction` but we don't need the "if" statement anymore :)
7. Task 10 is useless, the session can be accessed from twig so only a minor update of the related template.
8. For this specific task, I don't know... because I don't know why it's needed: if it's *only* to display/hide something in front, this is something that can be managed in Twig.

## Cool, but why do we want to do it?

Many reasons:

- if every part is in the right place, it's easier to debug, maintain and improve;
- mainly, our long-term aim is to dramatically improve the performance of all products pages;
- and to be more precise, we will remove a lot of useless/not optimized SQL queries;

> To keep this pull request readable, I'll start this plan in another one => https://github.com/PrestaShop/PrestaShop/pull/8710

mickaelandrieu commented:
Hello @Sethbass, if you don't mind I will take the pull request: I'm not easy with a branch called "develop" when I have also the rights to push force on the real "develop" branch :trollface:

So, to avoid a mistake I create a new branch with all our commits: thanks for your help :)

eternoendless commented:
Reviewed 1 of 3 files at r7, 9 of 9 files at r8.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[src/PrestaShopBundle/Controller/Admin/ProductController.php, line 139 at r5](https://reviewable.io:443/reviews/prestashop/prestashop/8677#-L3m1AYG-2ZVD041eN-7:-L40TuAFUSc7iySUT9dP:b-6lbcpa) ([raw file](https://github.com/prestashop/prestashop/blob/6aee5703fce41c831032c2d572f0cf0f841c4a61/src/PrestaShopBundle/Controller/Admin/ProductController.php#L139)):*
<details><summary><i>Previously, mickaelandrieu (Mickaël Andrieu) wrote…</i></summary><blockquote>

I dont see the code you refers to.
</blockquote></details>

Glad you found it :)

---

*[src/PrestaShopBundle/Controller/Admin/ProductController.php, line 850 at r8](https://reviewable.io:443/reviews/prestashop/prestashop/8677#-L40URv8UAeyib1EgaOm:-L40URv8UAeyib1EgaOn:b-hoa9rp) ([raw file](https://github.com/prestashop/prestashop/blob/44c4e45d35fb38a3e6f0376e45f77de8fd3b4de1/src/PrestaShopBundle/Controller/Admin/ProductController.php#L850)):*
> ```PHP
> 
>     /**
>      * @deprecated since 1.7.3.0, to be removed in 1.8 rely on CommonController::renderFieldAction
> ```

Actually, 1.7.4.0 since this PR is based on develop

---

*[src/PrestaShopBundle/Model/Product/AdminModelAdapter.php, line 529 at r5](https://reviewable.io:443/reviews/prestashop/prestashop/8677#-L3mRiWIZf1KQSJShBkH:-L40W7B_ouWC_gczVRZa:b-y7dxzt) ([raw file](https://github.com/prestashop/prestashop/blob/e651445a3436b69f4b2e1a45062b5081ecd4789e/src/PrestaShopBundle/Model/Product/AdminModelAdapter.php#L529)):*
<details><summary><i>Previously, eternoendless (Pablo Borowicz) wrote…</i></summary><blockquote>

Where did this method go?
</blockquote></details>

Still missing: `public function getAttributesResume()`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8677)*
<!-- Sent from Reviewable.io -->

mickaelandrieu commented:
> Still missing: public function getAttributesResume()

~~I don't know, it's not one of my commits~~: I've re-introduced it, nice spot!

> Actually, 1.7.4.0 since this PR is based on develop

true, I update.

mickaelandrieu committed:
Updated deprecation message

mickaelandrieu committed:
Reintroduced missing AdminModelAdapter:getAttributesResume function

eternoendless commented:
Reviewed 2 of 2 files at r9, 2 of 2 files at r10.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[src/Adapter/Product/FilterParametersUpdater.php, line 70 at r5](https://reviewable.io:443/reviews/prestashop/prestashop/8677#-L3lz27q708HwAXEZsL7:-L4plmhPX4HTfeR7qbqF:b-kw3mdx) ([raw file](https://github.com/prestashop/prestashop/blob/6aee5703fce41c831032c2d572f0cf0f841c4a61/src/Adapter/Product/FilterParametersUpdater.php#L70)):*
<details><summary><i>Previously, eternoendless (Pablo Borowicz) wrote…</i></summary><blockquote>

Passing parameters by reference is not generally not a good idea unless you have a very specific reason to do so.

I think this should be broken into 4 different methods like this:

```php
public function getOffset($offset, $filterParameters)
{
    return ($offset === 'last' && isset($filterParameters['last_offset'])) ? $filterParameters['last_offset'] : $offset;
}
```
</blockquote></details>

Still true

---

*[src/PrestaShopBundle/Model/Product/AdminModelAdapter.php, line 435 at r5](https://reviewable.io:443/reviews/prestashop/prestashop/8677#-L3mQWWiR74-sPPQN7t1:-L4plo6cICsPgPKjN9X4:b-mkiyol) ([raw file](https://github.com/prestashop/prestashop/blob/6aee5703fce41c831032c2d572f0cf0f841c4a61/src/PrestaShopBundle/Model/Product/AdminModelAdapter.php#L435)):*
<details><summary><i>Previously, eternoendless (Pablo Borowicz) wrote…</i></summary><blockquote>

IMO it's better to make assignments explicit instead of burying it in the method:

```php
$this->formData['step1'] = $this->getStep1FormData($product);
$this->formData['step2'] = $this->getStep2FormData($product);
// ...
```

I believe this way the code is easier to read
</blockquote></details>

Do you think you can make it like this?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8677)*
<!-- Sent from Reviewable.io -->

kpodemski commented on src/Core/Product/ProductCsvExporter.php:
```diff
@@ -0,0 +1,105 @@
+            'sav_quantity' => $this->trans('Quantity', 'Admin.Global'),
+            'badge_danger' => $this->trans('Status', 'Admin.Global'),
+            'position' => $this->trans('Position', 'Admin.Global'),
+        );
```
can we please have some hook here, and for the data to be able to export custom data?

mickaelandrieu replied:
it's an improvement I want to introduce but after the refactoring. As you can see, the split already allow us to add new useful features :+1:

Also, as it will be now a service you will be able to override it in every module to do anything you want ;)

mickaelandrieu commented:
> Still true

I know but I don't want to create new public methods that we must have to maintain when I'm not sure of the implementation.

You can see this pr as a first step: splitting and re ordering. But if I introduce new public methods this become part of our api and you wont allow me to remove it if I find a better way to do latter... what do you suggest?

eternoendless commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[src/Adapter/Product/FilterParametersUpdater.php, line 70 at r5](https://reviewable.io:443/reviews/prestashop/prestashop/8677#-L3lz27q708HwAXEZsL7:-L5ibDh02WR01fug4b6o:b-7g5mbc) ([raw file](https://github.com/prestashop/prestashop/blob/6aee5703fce41c831032c2d572f0cf0f841c4a61/src/Adapter/Product/FilterParametersUpdater.php#L70)):*
> I know but I don't want to create new public methods that we must have to maintain when I'm not sure of the implementation.
> 
> You can see this pr as a first step: splitting and re ordering. But if I introduce new public methods this become part of our api and you wont allow me to remove it if I find a better way to do latter... what do you suggest?

Same could be said of the currently proposed implementation: the pass-by-reference would also be part of the API. Either way, there are two things to consider:

1. As I believe you said once, classes on the `Adapter` namespace are to be considered transitional (aka **internal**) and shouldn't be relied upon by consumers.
2. I think it's better to make a getter and risk deprecating it, than to keep a bad interface for historical reasons. Remember the Boy Scout principle – "leave the camp cleaner than you found it".

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8677)*
<!-- Sent from Reviewable.io -->

mickaelandrieu commented:
@Sethbass I'm gonna close this one in favor of https://github.com/PrestaShop/PrestaShop/pull/8710.

Thanks for your help, this will - probably - be part of the next release.
